<a href="https://colab.research.google.com/github/nafalrifky/midterm-deep-learning/blob/main/midterm_Online_Transaction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Deep Learning - Online Transaction
## Nafal Rifky Atsilah Maulana - 1103223106

In this assignment, you will develop end-to-end machine learning and deep learning models using the provided dataset. This includes data preprocessing, model selection, training, evaluation, and comparison of performance across traditional machine learning models and deep learning architectures. Your task is to build a comprehensive pipeline and analyze the results to draw meaningful conclusions.


## Load Transaction Data




**Reasoning**:
The first step is to import the pandas library to enable data manipulation with DataFrames, as requested in the instructions.



In [29]:
import pandas as pd
print("pandas imported successfully.")

pandas imported successfully.


**Reasoning**:
Now that pandas is imported, I will load the `train_transaction.csv` dataset into a DataFrame named `df_train` using the file path provided in `corrected_file_paths`.



In [30]:
df_train = pd.read_csv(corrected_file_paths['train_transaction'])
print("train_transaction.csv loaded into df_train.")

train_transaction.csv loaded into df_train.


**Reasoning**:
Following the instructions, I will now load the `test_transaction.csv` dataset into a DataFrame named `df_test` using the file path provided in `corrected_file_paths`.



In [31]:
df_test = pd.read_csv(corrected_file_paths['test_transaction'])
print("test_transaction.csv loaded into df_test.")

test_transaction.csv loaded into df_test.


**Reasoning**:
Now that both `df_train` and `df_test` have been loaded, I will display the first few rows of `df_train` to inspect its structure and data, as per the instructions.



In [32]:
print("First 5 rows of df_train:")
df_train.head()

First 5 rows of df_train:


TransactionID  isFraud  TransactionDT  TransactionAmt ProductCD  card1  \
0        2987000        0          86400            68.5         W  13926   
1        2987001        0          86401            29.0         W   2755   
2        2987002        0          86469            59.0         W   4663   
3        2987003        0          86499            50.0         W  18132   
4        2987004        0          86506            50.0         H   4497   

   card2  card3       card4  card5  ... V330  V331  V332  V333  V334 V335  \
0    NaN  150.0    discover  142.0  ...  NaN   NaN   NaN   NaN   NaN  NaN   
1  404.0  150.0  mastercard  102.0  ...  NaN   NaN   NaN   NaN   NaN  NaN   
2  490.0  150.0        visa  166.0  ...  NaN   NaN   NaN   NaN   NaN  NaN   
3  567.0  150.0  mastercard  117.0  ...  NaN   NaN   NaN   NaN   NaN  NaN   
4  514.0  150.0  mastercard  102.0  ...  0.0   0.0   0.0   0.0   0.0  0.0   

  V336  V337  V338  V339  
0  NaN   NaN   NaN   NaN  
1  NaN   NaN   NaN   NaN  
2  NaN   NaN   NaN   NaN  
3  NaN   NaN   NaN   NaN  
4  0.0   0.0   0.0   0.0  

[5 rows x 394 columns]

**Reasoning**:
To fulfill the instruction to verify column names, non-null counts, and data types, I will now use the `.info()` method on `df_train`.



In [33]:
print("Information about df_train (columns, non-null counts, data types):")
df_train.info()

Information about df_train (columns, non-null counts, data types):
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 122029 entries, 0 to 122028
Columns: 394 entries, TransactionID to V339
dtypes: float64(376), int64(4), object(14)
memory usage: 366.8+ MB


**Reasoning**:
Now, to complete the subtask, I will verify the presence of the target column 'isFraud' in `df_train` by checking its column names.



In [34]:
if 'isFraud' in df_train.columns:
    print("'isFraud' column found in df_train.")
else:
    print("'isFraud' column NOT found in df_train.")

'isFraud' column found in df_train.


## Initial Data Preprocessing and Feature Engineering

### Subtask:
Lakukan pra-pemrosesan awal seperti imputasi nilai hilang untuk kolom `TransactionDT`, konversi ke format datetime, dan rekayasa fitur temporal (jam, hari dalam seminggu, hari dalam setahun). Juga, buat fitur agregat untuk 'card1' dan 'card4' berdasarkan 'TransactionAmt'.


**Reasoning**:
The first instruction is to handle missing values in the 'TransactionDT' column for both `df_train` and `df_test` by imputing them with the median. I will check for missing values, calculate the median, and then fill the missing values.



In [35]:
print("Checking and imputing missing values in 'TransactionDT'...")

# Impute missing values in df_train['TransactionDT']
if df_train['TransactionDT'].isnull().any():
    median_transaction_dt_train = df_train['TransactionDT'].median()
    df_train['TransactionDT'].fillna(median_transaction_dt_train, inplace=True)
    print(f"Missing values in df_train['TransactionDT'] imputed with median: {median_transaction_dt_train}")
else:
    print("No missing values found in df_train['TransactionDT']")

# Impute missing values in df_test['TransactionDT']
if df_test['TransactionDT'].isnull().any():
    median_transaction_dt_test = df_test['TransactionDT'].median()
    df_test['TransactionDT'].fillna(median_transaction_dt_test, inplace=True)
    print(f"Missing values in df_test['TransactionDT'] imputed with median: {median_transaction_dt_test}")
else:
    print("No missing values found in df_test['TransactionDT']")

Checking and imputing missing values in 'TransactionDT'...
No missing values found in df_train['TransactionDT']
No missing values found in df_test['TransactionDT']


**Reasoning**:
Now that missing values in 'TransactionDT' have been handled (or confirmed not to exist), the next step is to convert this column to datetime objects using the `START_DATE` variable as a reference. This will involve converting `START_DATE` to datetime and 'TransactionDT' to timedelta.



In [36]:
print("Converting 'TransactionDT' to datetime objects...")

# Convert START_DATE to datetime
start_date_datetime = pd.to_datetime(START_DATE)

# Convert 'TransactionDT' in df_train to datetime
df_train['TransactionDT'] = start_date_datetime + pd.to_timedelta(df_train['TransactionDT'], unit='s')
print("'TransactionDT' in df_train converted to datetime.")

# Convert 'TransactionDT' in df_test to datetime
df_test['TransactionDT'] = start_date_datetime + pd.to_timedelta(df_test['TransactionDT'], unit='s')
print("'TransactionDT' in df_test converted to datetime.")

Converting 'TransactionDT' to datetime objects...
'TransactionDT' in df_train converted to datetime.
'TransactionDT' in df_test converted to datetime.


**Reasoning**:
The next step is to extract temporal features such as hour, day of week, and day of year from the newly created 'TransactionDT' datetime column for both `df_train` and `df_test` as per the instructions.



In [37]:
print("Extracting temporal features from 'TransactionDT'...")

# Extract temporal features for df_train
df_train['Transaction_hour'] = df_train['TransactionDT'].dt.hour
df_train['Transaction_day_of_week'] = df_train['TransactionDT'].dt.dayofweek
df_train['Transaction_day_of_year'] = df_train['TransactionDT'].dt.dayofyear
print("Temporal features extracted for df_train.")

# Extract temporal features for df_test
df_test['Transaction_hour'] = df_test['TransactionDT'].dt.hour
df_test['Transaction_day_of_week'] = df_test['TransactionDT'].dt.dayofweek
df_test['Transaction_day_of_year'] = df_test['TransactionDT'].dt.dayofyear
print("Temporal features extracted for df_test.")

Extracting temporal features from 'TransactionDT'...
Temporal features extracted for df_train.
Temporal features extracted for df_test.


**Reasoning**:
As per the subtask instructions, I will now create aggregated features based on the 'card1' column for both `df_train` and `df_test`. This involves grouping by 'card1' and calculating the mean, standard deviation, and count of 'TransactionAmt'.



In [38]:
print("Creating aggregated features for 'card1'...")

# Aggregate features for df_train based on 'card1'
card1_agg_train = df_train.groupby('card1')['TransactionAmt'].agg(['mean', 'std', 'count']).reset_index()
card1_agg_train.columns = ['card1', 'card1_TransactionAmt_mean', 'card1_TransactionAmt_std', 'card1_TransactionAmt_count']
df_train = pd.merge(df_train, card1_agg_train, on='card1', how='left')
print("Aggregated features for 'card1' added to df_train.")

# Aggregate features for df_test based on 'card1'
card1_agg_test = df_test.groupby('card1')['TransactionAmt'].agg(['mean', 'std', 'count']).reset_index()
card1_agg_test.columns = ['card1', 'card1_TransactionAmt_mean', 'card1_TransactionAmt_std', 'card1_TransactionAmt_count']
df_test = pd.merge(df_test, card1_agg_test, on='card1', how='left')
print("Aggregated features for 'card1' added to df_test.")

Creating aggregated features for 'card1'...
Aggregated features for 'card1' added to df_train.
Aggregated features for 'card1' added to df_test.


**Reasoning**:
As per the subtask instructions, I will now create aggregated features based on the 'card4' column for both `df_train` and `df_test`. This involves grouping by 'card4' and calculating the mean, standard deviation, and count of 'TransactionAmt'.



In [39]:
print("Creating aggregated features for 'card4'...")

# Aggregate features for df_train based on 'card4'
card4_agg_train = df_train.groupby('card4')['TransactionAmt'].agg(['mean', 'std', 'count']).reset_index()
card4_agg_train.columns = ['card4', 'card4_TransactionAmt_mean', 'card4_TransactionAmt_std', 'card4_TransactionAmt_count']
df_train = pd.merge(df_train, card4_agg_train, on='card4', how='left')
print("Aggregated features for 'card4' added to df_train.")

# Aggregate features for df_test based on 'card4'
card4_agg_test = df_test.groupby('card4')['TransactionAmt'].agg(['mean', 'std', 'count']).reset_index()
card4_agg_test.columns = ['card4', 'card4_TransactionAmt_mean', 'card4_TransactionAmt_std', 'card4_TransactionAmt_count']
df_test = pd.merge(df_test, card4_agg_test, on='card4', how='left')
print("Aggregated features for 'card4' added to df_test.")

Creating aggregated features for 'card4'...
Aggregated features for 'card4' added to df_train.
Aggregated features for 'card4' added to df_test.


**Reasoning**:
As per the final instruction for this subtask, I need to display the first five rows of both `df_train` and `df_test` using `.head()` to inspect the newly created features. Following that, I will display the information summary using `.info()` for both DataFrames to verify data types and column existence.



In [40]:
print("First 5 rows of df_train with new features:")
display(df_train.head())

print("Information about df_train (columns, non-null counts, data types) with new features:")
df_train.info()

print("\nFirst 5 rows of df_test with new features:")
display(df_test.head())

print("Information about df_test (columns, non-null counts, data types) with new features:")
df_test.info()

First 5 rows of df_train with new features:


TransactionID  isFraud       TransactionDT  TransactionAmt ProductCD  \
0        2987000        0 2017-12-02 00:00:00            68.5         W   
1        2987001        0 2017-12-02 00:00:01            29.0         W   
2        2987002        0 2017-12-02 00:01:09            59.0         W   
3        2987003        0 2017-12-02 00:01:39            50.0         W   
4        2987004        0 2017-12-02 00:01:46            50.0         H   

   card1  card2  card3       card4  card5  ... V339  Transaction_hour  \
0  13926    NaN  150.0    discover  142.0  ...  NaN                 0   
1   2755  404.0  150.0  mastercard  102.0  ...  NaN                 0   
2   4663  490.0  150.0        visa  166.0  ...  NaN                 0   
3  18132  567.0  150.0  mastercard  117.0  ...  NaN                 0   
4   4497  514.0  150.0  mastercard  102.0  ...  0.0                 0   

   Transaction_day_of_week  Transaction_day_of_year  \
0                        5                      336   
1                        5                      336   
2                        5                      336   
3                        5                      336   
4                        5                      336   

   card1_TransactionAmt_mean card1_TransactionAmt_std  \
0                  96.500000                45.268643   
1                 219.485909               415.934329   
2                  87.849281                84.934209   
3                 112.150912               172.133289   
4                  40.000000                14.142136   

  card1_TransactionAmt_count  card4_TransactionAmt_mean  \
0                          7                 203.395295   
1                        154                 122.702411   
2                        153                 123.800858   
3                        746                 122.702411   
4                          2                 122.702411   

   card4_TransactionAmt_std  card4_TransactionAmt_count  
0                333.905949                      1794.0  
1                197.022774                     35607.0  
2                191.228857                     80393.0  
3                197.022774                     35607.0  
4                197.022774                     35607.0  

[5 rows x 403 columns]

Information about df_train (columns, non-null counts, data types) with new features:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 122029 entries, 0 to 122028
Columns: 403 entries, TransactionID to card4_TransactionAmt_count
dtypes: datetime64[ns](1), float64(381), int32(3), int64(4), object(14)
memory usage: 373.8+ MB

First 5 rows of df_test with new features:


TransactionID       TransactionDT  TransactionAmt ProductCD  card1  card2  \
0        3663549 2018-07-02 00:00:24           31.95         W  10409  111.0   
1        3663550 2018-07-02 00:01:03           49.00         W   4272  111.0   
2        3663551 2018-07-02 00:01:50          171.00         W   4476  574.0   
3        3663552 2018-07-02 00:01:50          284.95         W  10989  360.0   
4        3663553 2018-07-02 00:01:57           67.95         W  18018  452.0   

   card3       card4  card5  card6  ...  V339  Transaction_hour  \
0  150.0        visa  226.0  debit  ...   NaN                 0   
1  150.0        visa  226.0  debit  ...   NaN                 0   
2  150.0        visa  226.0  debit  ...   NaN                 0   
3  150.0        visa  166.0  debit  ...   NaN                 0   
4  150.0  mastercard  117.0  debit  ...   NaN                 0   

   Transaction_day_of_week  Transaction_day_of_year card1_TransactionAmt_mean  \
0                        0                      183                114.350909   
1                        0                      183                157.611694   
2                        0                      183                165.715000   
3                        0                      183                 93.437782   
4                        0                      183                162.086896   

  card1_TransactionAmt_std  card1_TransactionAmt_count  \
0               150.415884                          22   
1               331.341043                         235   
2               100.181422                          10   
3               115.227758                         239   
4               368.429429                         318   

   card4_TransactionAmt_mean  card4_TransactionAmt_std  \
0                 140.926026                254.200343   
1                 140.926026                254.200343   
2                 140.926026                254.200343   
3                 140.926026                254.200343   
4                 138.547560                255.184939   

   card4_TransactionAmt_count  
0                     82170.0  
1                     82170.0  
2                     82170.0  
3                     82170.0  
4                     39751.0  

[5 rows x 402 columns]

Information about df_test (columns, non-null counts, data types) with new features:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 125110 entries, 0 to 125109
Columns: 402 entries, TransactionID to card4_TransactionAmt_count
dtypes: datetime64[ns](1), float64(381), int32(3), int64(3), object(14)
memory usage: 382.3+ MB


## Identify Column Types and Handle Missing Values

### Subtask:
Identifikasi kembali kolom numerik dan kategorikal setelah rekayasa fitur. Imputasi nilai hilang di kolom numerik menggunakan median/mean dan di kolom kategorikal dengan 'Missing'.


**Reasoning**:
The first instruction is to separate columns into numerical and categorical lists for both `df_train` and `df_test`, excluding 'TransactionID' and 'isFraud'. I will iterate through the columns and assign them based on their data types.



In [41]:
print("Separating columns into numerical and categorical lists...")

# Identify numerical and categorical columns for df_train
numerical_cols_train = df_train.select_dtypes(include=['int64', 'float64', 'int32']).columns.tolist()
categorical_cols_train = df_train.select_dtypes(include=['object']).columns.tolist()

# Remove 'TransactionID' and 'isFraud' from the lists if present
if 'TransactionID' in numerical_cols_train:
    numerical_cols_train.remove('TransactionID')
if 'isFraud' in numerical_cols_train:
    numerical_cols_train.remove('isFraud')

# Identify numerical and categorical columns for df_test
numerical_cols_test = df_test.select_dtypes(include=['int64', 'float64', 'int32']).columns.tolist()
categorical_cols_test = df_test.select_dtypes(include=['object']).columns.tolist()

# Remove 'TransactionID' from the lists if present
if 'TransactionID' in numerical_cols_test:
    numerical_cols_test.remove('TransactionID')

print(f"Numerical columns (train): {len(numerical_cols_train)}")
print(f"Categorical columns (train): {len(categorical_cols_train)}")
print(f"Numerical columns (test): {len(numerical_cols_test)}")
print(f"Categorical columns (test): {len(categorical_cols_test)}")

Separating columns into numerical and categorical lists...
Numerical columns (train): 386
Categorical columns (train): 14
Numerical columns (test): 386
Categorical columns (test): 14


**Reasoning**:
Now that the numerical and categorical columns have been identified, I will proceed with imputing missing values in the numerical columns. As per the instructions, I will calculate the median for each numerical column from `df_train` and use it to fill missing values in both `df_train` and `df_test` for that column.



In [42]:
print("Imputing missing values in numerical columns...")

for col in numerical_cols_train:
    if df_train[col].isnull().any():
        median_val = df_train[col].median()
        df_train[col].fillna(median_val, inplace=True)
        df_test[col].fillna(median_val, inplace=True)
print("Missing values in numerical columns imputed using median from df_train for both dataframes.")

Imputing missing values in numerical columns...


/tmp/ipython-input-1487949001.py:6: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_train[col].fillna(median_val, inplace=True)
/tmp/ipython-input-1487949001.py:7: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try usin

Missing values in numerical columns imputed using median from df_train for both dataframes.


**Reasoning**:
The previous step successfully imputed numerical missing values, but a `FutureWarning` was raised about `inplace=True`. While not an error, it's good practice to update the code to avoid future issues. The next instruction is to impute missing values in categorical columns. I will iterate through the identified categorical columns and fill `NaN` values with the string 'Missing' for both `df_train` and `df_test`.



In [43]:
print("Imputing missing values in categorical columns...")

for col in categorical_cols_train:
    if df_train[col].isnull().any():
        df_train[col] = df_train[col].fillna('Missing')
    if df_test[col].isnull().any():
        df_test[col] = df_test[col].fillna('Missing')
print("Missing values in categorical columns imputed with 'Missing' for both dataframes.")

Imputing missing values in categorical columns...
Missing values in categorical columns imputed with 'Missing' for both dataframes.


**Reasoning**:
Now that both numerical and categorical missing values have been imputed, I need to verify that there are no remaining missing values in either `df_train` or `df_test` by printing the sum of nulls for both dataframes, as per the instructions.



In [44]:
print("Verifying no remaining missing values...")

missing_train = df_train.isnull().sum().sum()
missing_test = df_test.isnull().sum().sum()

print(f"Total missing values in df_train: {missing_train}")
print(f"Total missing values in df_test: {missing_test}")

if missing_train == 0 and missing_test == 0:
    print("No missing values found in df_train and df_test after imputation.")
else:
    print("WARNING: Missing values still exist in df_train or df_test.")

Verifying no remaining missing values...
Total missing values in df_train: 0
Total missing values in df_test: 550


**Reasoning**:
The previous verification step indicated that `df_test` still has 550 missing values. To address this and properly complete the subtask, I need to identify which specific columns in `df_test` still contain missing values. This will help understand if they are columns that were supposed to be imputed or if there's an issue with the column identification.



In [45]:
print("Identifying columns in df_test with remaining missing values...")

missing_cols_df_test = df_test.isnull().sum()
missing_cols_df_test = missing_cols_df_test[missing_cols_df_test > 0]

if not missing_cols_df_test.empty:
    print("Columns in df_test with remaining missing values and their counts:")
    print(missing_cols_df_test)
else:
    print("No missing values found in df_test.")

Identifying columns in df_test with remaining missing values...
Columns in df_test with remaining missing values and their counts:
D1    550
dtype: int64


**Reasoning**:
The previous numerical imputation step missed filling missing values in 'D1' of `df_test` because the imputation logic was conditional on `df_train` having missing values in that column. To properly address all missing numerical values as per the instructions (which state to use the median from `df_train` to fill `NaN`s in both `df_train` and `df_test`), I will re-execute the numerical imputation step, ensuring that the median from `df_train` is calculated for each numerical column and then used to fill missing values in *both* `df_train` and `df_test`, regardless of whether `df_train` itself has missing values for that column. This will ensure 'D1' in `df_test` is also handled.



In [46]:
print("Re-imputing missing values in numerical columns to ensure completeness...")

for col in numerical_cols_train:
    # Calculate median from df_train for the current column
    median_val = df_train[col].median()

    # Fill missing values in df_train
    if df_train[col].isnull().any():
        df_train[col] = df_train[col].fillna(median_val)

    # Fill missing values in df_test using the median from df_train
    if df_test[col].isnull().any():
        df_test[col] = df_test[col].fillna(median_val)

print("Missing values in numerical columns re-imputed correctly using median from df_train for both dataframes.")

Re-imputing missing values in numerical columns to ensure completeness...
Missing values in numerical columns re-imputed correctly using median from df_train for both dataframes.


**Reasoning**:
After re-imputing the numerical columns, I need to re-verify that there are no remaining missing values in either `df_train` or `df_test` by printing the sum of nulls for both dataframes, as per instruction #4.



In [47]:
print("Verifying no remaining missing values after re-imputation...")

missing_train = df_train.isnull().sum().sum()
missing_test = df_test.isnull().sum().sum()

print(f"Total missing values in df_train: {missing_train}")
print(f"Total missing values in df_test: {missing_test}")

if missing_train == 0 and missing_test == 0:
    print("No missing values found in df_train and df_test after re-imputation.")
else:
    print("WARNING: Missing values still exist in df_train or df_test.")

Verifying no remaining missing values after re-imputation...
Total missing values in df_train: 0
Total missing values in df_test: 0
No missing values found in df_train and df_test after re-imputation.


**Reasoning**:
Now that all missing values have been successfully imputed and verified, the final step for this subtask is to display the information summary (`.info()`) for both `df_train` and `df_test`. This will show the updated non-null counts and data types, confirming the imputation was successful and reflecting the current state of the dataframes.



In [48]:
print("Displaying info for df_train after imputation:")
df_train.info()

print("\nDisplaying info for df_test after imputation:")
df_test.info()

Displaying info for df_train after imputation:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 122029 entries, 0 to 122028
Columns: 403 entries, TransactionID to card4_TransactionAmt_count
dtypes: datetime64[ns](1), float64(381), int32(3), int64(4), object(14)
memory usage: 373.8+ MB

Displaying info for df_test after imputation:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 125110 entries, 0 to 125109
Columns: 402 entries, TransactionID to card4_TransactionAmt_count
dtypes: datetime64[ns](1), float64(381), int32(3), int64(3), object(14)
memory usage: 382.3+ MB


## One-Hot Encode Categorical Features

### Subtask:
Terapkan One-Hot Encoding pada semua kolom kategorikal di kedua DataFrame (`df_train` dan `df_test`). Pastikan konsistensi kolom antara data latih dan uji setelah encoding untuk menghindari perbedaan jumlah fitur.


**Reasoning**:
The first step, identifying categorical columns, has already been completed in a previous subtask where `categorical_cols_train` and `categorical_cols_test` were created. Now, as per instruction 2, I will apply One-Hot Encoding to the categorical columns in `df_train` using `pd.get_dummies()`.



In [49]:
print("Applying One-Hot Encoding to categorical columns in df_train...")

df_train = pd.get_dummies(df_train, columns=categorical_cols_train, dummy_na=False)

print("One-Hot Encoding applied to df_train.")

Applying One-Hot Encoding to categorical columns in df_train...
One-Hot Encoding applied to df_train.


**Reasoning**:
Following the application of One-Hot Encoding to `df_train`, I will now apply One-Hot Encoding to the categorical columns in `df_test` using `pd.get_dummies()`, as specified in instruction 3 of the subtask.



In [50]:
print("Applying One-Hot Encoding to categorical columns in df_test...")

df_test = pd.get_dummies(df_test, columns=categorical_cols_test, dummy_na=False)

print("One-Hot Encoding applied to df_test.")

Applying One-Hot Encoding to categorical columns in df_test...
One-Hot Encoding applied to df_test.


**Reasoning**:
Now that both `df_train` and `df_test` have been one-hot encoded, the next step is to identify all columns present in both DataFrames, as per instruction 4. This is a preparatory step for ensuring column consistency.



In [51]:
print("Identifying all columns in df_train and df_test...")

all_cols_train = set(df_train.columns)
all_cols_test = set(df_test.columns)

print(f"Number of columns in df_train: {len(all_cols_train)}")
print(f"Number of columns in df_test: {len(all_cols_test)}")

Identifying all columns in df_train and df_test...
Number of columns in df_train: 552
Number of columns in df_test: 547


**Reasoning**:
As per instruction 5, I will create a consistent list of columns for both DataFrames by taking the union of the columns from `df_train` and `df_test`. This ensures that both dataframes will have the same set of columns before further processing.



In [52]:
print("Creating a consistent column list for both df_train and df_test...")

consistent_cols = list(all_cols_train.union(all_cols_test))

print(f"Number of consistent columns: {len(consistent_cols)}")

Creating a consistent column list for both df_train and df_test...
Number of consistent columns: 553


**Reasoning**:
As per instruction 6, I will reindex both `df_train` and `df_test` to the `consistent_cols` list. Any columns present in `consistent_cols` but not originally in `df_train` or `df_test` will be added and filled with 0, ensuring consistency across both DataFrames.



In [53]:
print("Reindexing df_train and df_test to consistent columns...")

df_train = df_train.reindex(columns=consistent_cols, fill_value=0)
df_test = df_test.reindex(columns=consistent_cols, fill_value=0)

print("DataFrames reindexed. Number of columns in df_train:", len(df_train.columns))
print("DataFrames reindexed. Number of columns in df_test:", len(df_test.columns))

Reindexing df_train and df_test to consistent columns...
DataFrames reindexed. Number of columns in df_train: 553
DataFrames reindexed. Number of columns in df_test: 553


**Reasoning**:
As per instruction 7, I will now display the first five rows of both `df_train` and `df_test` using `.head()` to inspect the DataFrames after one-hot encoding and reindexing.



In [54]:
print("First 5 rows of df_train after one-hot encoding and reindexing:")
display(df_train.head())

print("\nFirst 5 rows of df_test after one-hot encoding and reindexing:")
display(df_test.head())

First 5 rows of df_train after one-hot encoding and reindexing:


V21  V309  V284  V98  V323  R_emaildomain_embarqmail.com  \
0  0.0   0.0   0.0  0.0   0.0                         False   
1  0.0   0.0   0.0  0.0   0.0                         False   
2  0.0   0.0   0.0  0.0   0.0                         False   
3  0.0   0.0   0.0  0.0   0.0                         False   
4  0.0   0.0   0.0  0.0   0.0                         False   

   P_emaildomain_verizon.net  V118  V254  V220  ...   M7_F  V333  V321  V290  \
0                      False   1.0   1.0   0.0  ...  False   0.0   0.0   1.0   
1                      False   1.0   1.0   0.0  ...  False   0.0   0.0   1.0   
2                      False   1.0   1.0   0.0  ...   True   0.0   0.0   1.0   
3                      False   1.0   1.0   0.0  ...  False   0.0   0.0   1.0   
4                      False   1.0   1.0   0.0  ...  False   0.0   0.0   1.0   

   M3_Missing  V293  R_emaildomain_frontiernet.net  V188  V256  V316  
0       False   0.0                          False   1.0   1.0   0.0  
1        True   0.0                          False   1.0   1.0   0.0  
2       False   0.0                          False   1.0   1.0   0.0  
3        True   1.0                          False   1.0   1.0  50.0  
4        True   0.0                          False   1.0   1.0   0.0  

[5 rows x 553 columns]


First 5 rows of df_test after one-hot encoding and reindexing:


V21       V309  V284  V98  V323  R_emaildomain_embarqmail.com  \
0  0.0   0.000000   0.0  0.0   0.0                         False   
1  0.0   0.000000   0.0  0.0   0.0                         False   
2  0.0   0.000000   0.0  0.0   0.0                         False   
3  0.0   0.000000   0.0  0.0   0.0                         False   
4  0.0  67.949997   1.0  1.0   0.0                         False   

   P_emaildomain_verizon.net  V118  V254  V220  ...   M7_F  V333  V321  V290  \
0                      False   1.0   1.0   0.0  ...  False   0.0   0.0   1.0   
1                      False   1.0   1.0   0.0  ...  False   0.0   0.0   1.0   
2                      False   1.0   1.0   0.0  ...   True   0.0   0.0   1.0   
3                      False   1.0   1.0   0.0  ...  False   0.0   0.0   1.0   
4                      False   1.0   1.0   0.0  ...   True   0.0   0.0   1.0   

   M3_Missing  V293  R_emaildomain_frontiernet.net  V188  V256  V316  
0       False   0.0                          False   1.0   1.0   0.0  
1       False   0.0                          False   1.0   1.0   0.0  
2       False   0.0                          False   1.0   1.0   0.0  
3       False   0.0                          False   1.0   1.0   0.0  
4       False   0.0                          False   1.0   1.0   0.0  

[5 rows x 553 columns]

**Reasoning**:
Now that the head of both dataframes has been displayed, I will display the information summary (`.info()`) for both `df_train` and `df_test` to verify data types and the final count of columns, as per instruction 8.



In [55]:
print("Displaying info for df_train after one-hot encoding and reindexing:")
df_train.info()

print("\nDisplaying info for df_test after one-hot encoding and reindexing:")
df_test.info()

Displaying info for df_train after one-hot encoding and reindexing:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 122029 entries, 0 to 122028
Columns: 553 entries, V21 to V316
dtypes: bool(163), datetime64[ns](1), float64(381), int32(3), int64(5)
memory usage: 380.7 MB

Displaying info for df_test after one-hot encoding and reindexing:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 125110 entries, 0 to 125109
Columns: 553 entries, V21 to V316
dtypes: bool(159), datetime64[ns](1), float64(381), int32(3), int64(9)
memory usage: 393.6 MB


## Separate Target and Features, Train-Validation Split

### Subtask:
Pisahkan variabel target (`isFraud`) dari fitur (`X`) di `df_train`. Kemudian, bagi `X` dan `y` menjadi set pelatihan dan validasi (`X_train`, `X_val`, `y_train`, `y_val`) menggunakan stratifikasi untuk menjaga proporsi kelas penipuan.


**Reasoning**:
The first instruction is to import the `train_test_split` function from `sklearn.model_selection` to enable the splitting of data into training and validation sets.



In [56]:
from sklearn.model_selection import train_test_split
print("train_test_split imported successfully.")

train_test_split imported successfully.


**Reasoning**:
Following the instructions, I will separate the target variable 'isFraud' from `df_train` and store it in a new Series named `y`.



In [57]:
print("Separating target variable 'isFraud' into y...")
y = df_train['isFraud']
print("Target variable 'isFraud' separated into y.")

Separating target variable 'isFraud' into y...
Target variable 'isFraud' separated into y.


**Reasoning**:
As per the instructions, I will now create the feature set `X` by dropping the specified columns ('isFraud', 'TransactionID', and 'TransactionDT') from `df_train`.



In [58]:
print("Creating feature set X by dropping 'isFraud', 'TransactionID', and 'TransactionDT' from df_train...")
X = df_train.drop(['isFraud', 'TransactionID', 'TransactionDT'], axis=1)
print("Feature set X created.")

Creating feature set X by dropping 'isFraud', 'TransactionID', and 'TransactionDT' from df_train...
Feature set X created.


**Reasoning**:
Now that the target variable `y` and feature set `X` have been prepared, I will split them into training and validation sets as instructed, ensuring stratification to maintain the proportion of fraudulent transactions.



In [59]:
print("Splitting X and y into training and validation sets...")
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)
print("Data split into X_train, X_val, y_train, y_val.")

Splitting X and y into training and validation sets...
Data split into X_train, X_val, y_train, y_val.


**Reasoning**:
As per the final instruction for this subtask, I will print the shapes of `X_train`, `X_val`, `y_train`, and `y_val` to verify that the split was performed correctly and to check the dimensions of the resulting datasets.



In [60]:
print("Verifying shapes of training and validation sets:")
print(f"X_train shape: {X_train.shape}")
print(f"X_val shape: {X_val.shape}")
print(f"y_train shape: {y_train.shape}")
print(f"y_val shape: {y_val.shape}")

Verifying shapes of training and validation sets:
X_train shape: (97623, 550)
X_val shape: (24406, 550)
y_train shape: (97623,)
y_val shape: (24406,)


## Handle Class Imbalance with SMOTE

### Subtask:
Terapkan teknik Synthetic Minority Over-sampling Technique (SMOTE) pada data pelatihan (`X_train`, `y_train`) untuk mengatasi ketidakseimbangan kelas dalam deteksi penipuan, menghasilkan `X_resampled` dan `y_resampled`.


**Reasoning**:
The first step is to import the `SMOTE` class from the `imblearn.over_sampling` library to enable the application of the Synthetic Minority Over-sampling Technique.



In [61]:
from imblearn.over_sampling import SMOTE
print("SMOTE imported successfully.")

SMOTE imported successfully.


**Reasoning**:
As per the instructions, I will initialize the SMOTE object with `random_state=42` and then apply it to `X_train` and `y_train` using the `.fit_resample()` method to generate the resampled datasets `X_resampled` and `y_resampled`.



In [62]:
print("Applying SMOTE to X_train and y_train...")

smote = SMOTE(random_state=42)
X_resampled, y_resampled = smote.fit_resample(X_train, y_train)

print("SMOTE applied successfully.")

Applying SMOTE to X_train and y_train...
SMOTE applied successfully.


**Reasoning**:
As per the instructions, I will now print the shapes of `X_resampled` and `y_resampled` to verify the number of samples after applying SMOTE.



In [63]:
print("Verifying shapes of resampled datasets:")
print(f"X_resampled shape: {X_resampled.shape}")
print(f"y_resampled shape: {y_resampled.shape}")

Verifying shapes of resampled datasets:
X_resampled shape: (190562, 550)
y_resampled shape: (190562,)


**Reasoning**:
As per the final instruction for this subtask, I will print the value counts of the target class in `y_resampled` to verify that the class imbalance has been addressed.



In [64]:
print("Verifying target class distribution in y_resampled:")
print(y_resampled.value_counts())

Verifying target class distribution in y_resampled:
isFraud
0    95281
1    95281
Name: count, dtype: int64


## Train Traditional Machine Learning Model (Classification)

### Subtask:
Pilih dan latih model machine learning tradisional yang sesuai untuk klasifikasi (misalnya, RandomForestClassifier atau LightGBM Classifier) menggunakan `X_resampled` dan `y_resampled`. Latih model untuk memprediksi `isFraud`.


**Reasoning**:
The first instruction is to import the `RandomForestClassifier` from `sklearn.ensemble` to prepare for model selection and training.



In [65]:
from sklearn.ensemble import RandomForestClassifier
print("RandomForestClassifier imported successfully.")

RandomForestClassifier imported successfully.


**Reasoning**:
As per the instructions, I will initialize the RandomForestClassifier with `random_state=42` to ensure reproducibility, and store the instance in a variable named `model_traditional`.



In [66]:
print("Initializing RandomForestClassifier model...")

model_traditional = RandomForestClassifier(random_state=42)

print("RandomForestClassifier initialized successfully.")

Initializing RandomForestClassifier model...
RandomForestClassifier initialized successfully.


**Reasoning**:
Now that the `RandomForestClassifier` has been initialized, the next step is to train it using the resampled training data (`X_resampled` and `y_resampled`), as specified in instruction 3 of the subtask. This will involve calling the `.fit()` method on the `model_traditional` instance.



In [67]:
print("Training the RandomForestClassifier model...")

model_traditional.fit(X_resampled, y_resampled)

print("RandomForestClassifier model trained successfully.")

Training the RandomForestClassifier model...
RandomForestClassifier model trained successfully.


## Evaluate Traditional ML Model

### Subtask:
Evaluasi performa model machine learning tradisional pada set validasi (`X_val`, `y_val`) menggunakan metrik klasifikasi seperti Akurasi, Presisi, Recall, F1-Score, dan AUC-ROC. Tampilkan Confusion Matrix.


**Reasoning**:
The first instruction is to use the trained traditional machine learning model (`model_traditional`) to make class predictions (`y_pred_traditional`) on the `X_val` dataset. This will provide the predicted labels for evaluating classification metrics.



In [68]:
print("Making class predictions on the validation set...")
y_pred_traditional = model_traditional.predict(X_val)
print("Class predictions (y_pred_traditional) made successfully.")

Making class predictions on the validation set...
Class predictions (y_pred_traditional) made successfully.


**Reasoning**:
As per the instructions, I will now create probability predictions for the positive class (isFraud=1) on the `X_val` dataset using the `.predict_proba()` method of the trained traditional model. This is necessary for calculating the AUC-ROC score.



In [69]:
print("Making probability predictions on the validation set...")
y_proba_traditional = model_traditional.predict_proba(X_val)[:, 1] # Get probabilities for the positive class (isFraud=1)
print("Probability predictions (y_proba_traditional) made successfully.")

Making probability predictions on the validation set...
Probability predictions (y_proba_traditional) made successfully.


**Reasoning**:
As per instruction 3, I need to import the required evaluation metrics from `sklearn.metrics` to calculate the performance of the traditional model.



In [70]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score, confusion_matrix
print("Required evaluation metrics imported successfully.")

Required evaluation metrics imported successfully.


**Reasoning**:
Now that the evaluation metrics have been imported, I will calculate and print the accuracy, precision, recall, F1-score, and AUC-ROC using `y_val`, `y_pred_traditional`, and `y_proba_traditional`, as specified in instruction 4.



In [71]:
print("Calculating and printing evaluation metrics...")

accuracy = accuracy_score(y_val, y_pred_traditional)
precision = precision_score(y_val, y_pred_traditional)
recall = recall_score(y_val, y_pred_traditional)
f1 = f1_score(y_val, y_pred_traditional)
roc_auc = roc_auc_score(y_val, y_proba_traditional)

print(f"Accuracy: {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1-Score: {f1:.4f}")
print(f"AUC-ROC: {roc_auc:.4f}")

Calculating and printing evaluation metrics...
Accuracy: 0.9849
Precision: 0.9252
Recall: 0.4017
F1-Score: 0.5602
AUC-ROC: 0.9072


**Reasoning**:
As per instruction 5, I will now print the confusion matrix using `confusion_matrix(y_val, y_pred_traditional)` to visualize the performance of the traditional model.



In [72]:
print("Printing Confusion Matrix:")
conf_matrix = confusion_matrix(y_val, y_pred_traditional)
print(conf_matrix)

Printing Confusion Matrix:
[[23802    19]
 [  350   235]]


## Prepare Data for Deep Learning (Classification)

### Subtask:
Skalakan fitur numerik pada `X_resampled` dan `X_val` menggunakan `MinMaxScaler` untuk mempersiapkan data bagi model deep learning.


**Reasoning**:
The first instruction is to import the `MinMaxScaler` from `sklearn.preprocessing` to prepare for scaling the numerical features.



In [73]:
from sklearn.preprocessing import MinMaxScaler
print("MinMaxScaler imported successfully.")

MinMaxScaler imported successfully.


**Reasoning**:
As per the instructions, I will initialize a `MinMaxScaler` object. It's good practice to assign it to a variable, usually named `scaler`.



In [74]:
print("Initializing MinMaxScaler object...")

scaler = MinMaxScaler()

print("MinMaxScaler object initialized successfully.")

Initializing MinMaxScaler object...
MinMaxScaler object initialized successfully.


**Reasoning**:
As per instruction 3, I will now fit the `MinMaxScaler` on the numerical columns of `X_resampled` and then transform these columns, storing the result in `X_resampled_scaled`. I will also transform the corresponding numerical columns in `X_val` using the same fitted scaler, storing the result in `X_val_scaled`, as per instruction 4. I need to explicitly select the numerical columns that are not boolean for scaling.



In [75]:
print("Fitting scaler on X_resampled and transforming X_resampled and X_val...")

# Identify numerical columns in X_resampled that are not boolean for scaling
numerical_cols_for_scaling = X_resampled.select_dtypes(include=['int64', 'float64', 'int32']).columns.tolist()

# Fit the scaler on the numerical columns of X_resampled and transform
X_resampled_scaled_array = scaler.fit_transform(X_resampled[numerical_cols_for_scaling])

# Transform the numerical columns of X_val using the same fitted scaler
X_val_scaled_array = scaler.transform(X_val[numerical_cols_for_scaling])

# Create new DataFrames with scaled numerical columns, keeping other columns (e.g., boolean) as they are
# First, copy the original DataFrames
X_resampled_scaled = X_resampled.copy()
X_val_scaled = X_val.copy()

# Then update the numerical columns with their scaled versions
X_resampled_scaled[numerical_cols_for_scaling] = X_resampled_scaled_array
X_val_scaled[numerical_cols_for_scaling] = X_val_scaled_array

print("X_resampled_scaled and X_val_scaled created with numerical features scaled.")

Fitting scaler on X_resampled and transforming X_resampled and X_val...
X_resampled_scaled and X_val_scaled created with numerical features scaled.


**Reasoning**:
As per instruction 5, I will print the shapes of `X_resampled_scaled` and `X_val_scaled` to verify that the scaling and transformation were applied correctly and to confirm the dimensions of the resulting datasets.



In [76]:
print("Verifying shapes of scaled datasets:")
print(f"X_resampled_scaled shape: {X_resampled_scaled.shape}")
print(f"X_val_scaled shape: {X_val_scaled.shape}")

Verifying shapes of scaled datasets:
X_resampled_scaled shape: (190562, 550)
X_val_scaled shape: (24406, 550)


**Reasoning**:
As per instruction 6, I will now display the first 5 rows of `X_resampled_scaled` and `X_val_scaled` using `.head()` to inspect the scaled data.



In [77]:
print("First 5 rows of X_resampled_scaled:")
display(X_resampled_scaled.head())

print("\nFirst 5 rows of X_val_scaled:")
display(X_val_scaled.head())

First 5 rows of X_resampled_scaled:


V21  V309  V284  V98      V323  R_emaildomain_embarqmail.com  \
0  0.0   0.0   0.0  0.0  0.000000                         False   
1  0.0   0.0   0.0  0.0  0.000000                         False   
2  0.0   0.0   0.0  0.0  0.000000                         False   
3  0.0   0.0   0.0  0.0  0.022727                         False   
4  0.0   0.0   0.0  0.0  0.022727                         False   

   P_emaildomain_verizon.net  V118  V254  V220  ...   M7_F      V333  V321  \
0                      False   0.0   0.0   0.0  ...  False  0.000000   0.0   
1                      False   0.0   0.0   0.0  ...  False  0.000000   0.0   
2                      False   0.0   0.0   0.0  ...  False  0.000000   0.0   
3                      False   0.0   0.0   0.0  ...  False  0.007059   0.0   
4                      False   0.0   0.0   0.0  ...  False  0.009412   0.0   

   V290  M3_Missing      V293  R_emaildomain_frontiernet.net      V188  \
0   0.0        True  0.000000                          False  0.058824   
1   0.0        True  0.000000                          False  0.058824   
2   0.0        True  0.000000                          False  0.058824   
3   0.0        True  0.015385                          False  0.058824   
4   0.0        True  0.000000                          False  0.058824   

       V256    V316  
0  0.012195  0.0000  
1  0.012195  0.0000  
2  0.012195  0.0000  
3  0.012195  0.0125  
4  0.000000  0.0000  

[5 rows x 550 columns]


First 5 rows of X_val_scaled:


V21  V309  V284  V98  V323  R_emaildomain_embarqmail.com  \
20721  0.0   0.0   0.0  0.0   0.0                         False   
64129  0.0   0.0   0.0  0.0   0.0                         False   
97493  0.0   0.0   0.0  0.0   0.0                         False   
92274  0.0   0.0   0.0  0.0   0.0                         False   
8002   0.0   0.0   0.0  0.0   0.0                         False   

       P_emaildomain_verizon.net  V118  V254  V220  ...   M7_F  V333  V321  \
20721                      False   0.0   0.0   0.0  ...  False   0.0   0.0   
64129                      False   0.0   0.0   0.0  ...  False   0.0   0.0   
97493                      False   0.0   0.0   0.0  ...  False   0.0   0.0   
92274                      False   0.0   0.0   0.0  ...  False   0.0   0.0   
8002                       False   0.0   0.0   0.0  ...  False   0.0   0.0   

       V290  M3_Missing  V293  R_emaildomain_frontiernet.net      V188  \
20721   0.0        True   0.0                          False  0.058824   
64129   0.0        True   0.0                          False  0.058824   
97493   0.0        True   0.0                          False  0.058824   
92274   0.0        True   0.0                          False  0.058824   
8002    0.0       False   0.0                          False  0.058824   

           V256  V316  
20721  0.012195   0.0  
64129  0.012195   0.0  
97493  0.012195   0.0  
92274  0.012195   0.0  
8002   0.012195   0.0  

[5 rows x 550 columns]

## Train Deep Learning Model (Classification)

### Subtask:
Bangun dan latih model neural network sederhana menggunakan TensorFlow/Keras untuk tugas klasifikasi biner. Gunakan `X_resampled` yang sudah diskalakan dan `y_resampled` untuk pelatihan.


**Reasoning**:
The first instruction is to import the necessary libraries, `tensorflow` and modules like `Sequential` and `Dense` from `keras`, to begin building the deep learning model.



In [78]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

print("TensorFlow and Keras modules imported successfully.")

TensorFlow and Keras modules imported successfully.


**Reasoning**:
As per instruction 2, I will initialize a Sequential model. This is the foundation for building a layered neural network in Keras.



In [79]:
print("Initializing the Sequential model...")

model_dl = Sequential()

print("Sequential model initialized successfully.")

Initializing the Sequential model...
Sequential model initialized successfully.


**Reasoning**:
As per instruction 3, I will add the input layer to the sequential model. This layer will be a `Dense` layer with a specified number of units (e.g., 128), `relu` activation, and `input_dim` set to the number of features in `X_resampled_scaled`.



In [80]:
print("Adding input layer to the model...")

# Get the number of features from X_resampled_scaled
input_dim = X_resampled_scaled.shape[1]

model_dl.add(Dense(128, activation='relu', input_dim=input_dim))

print("Input layer added successfully.")

Adding input layer to the model...
Input layer added successfully.


/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


**Reasoning**:
Following the instructions, I will now add a hidden layer to the neural network model. This will be a `Dense` layer with a `relu` activation function, building upon the input layer already defined.



In [81]:
print("Adding hidden layer to the model...")

model_dl.add(Dense(64, activation='relu')) # Adding a hidden layer with 64 units

print("Hidden layer added successfully.")

Adding hidden layer to the model...
Hidden layer added successfully.


**Reasoning**:
As per instruction 5, I will now add the output layer to the model. For binary classification, this layer will be a `Dense` layer with 1 unit and a `sigmoid` activation function.



In [82]:
print("Adding output layer to the model...")

model_dl.add(Dense(1, activation='sigmoid')) # Output layer for binary classification

print("Output layer added successfully.")

Adding output layer to the model...
Output layer added successfully.


**Reasoning**:
As per instruction 6, I will now compile the deep learning model. This involves specifying the `Adam` optimizer, `binary_crossentropy` as the loss function (suitable for binary classification), and `accuracy` as the metric to monitor during training.



In [83]:
print("Compiling the deep learning model...")

model_dl.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

print("Model compiled successfully.")

Compiling the deep learning model...
Model compiled successfully.


**Reasoning**:
As per instruction 7, I will now train the compiled deep learning model using `X_resampled_scaled` and `y_resampled` for training, and `(X_val_scaled, y_val)` for validation, with specified epochs and batch size. This will initiate the learning process.



In [84]:
print("Training the deep learning model...")

history = model_dl.fit(X_resampled_scaled, y_resampled,
                        epochs=10,
                        batch_size=32,
                        validation_data=(X_val_scaled, y_val),
                        verbose=1)

print("Deep learning model trained successfully.")

Training the deep learning model...
Epoch 1/10
5956/5956 ━━━━━━━━━━━━━━━━━━━━ 27s 4ms/step - accuracy: 0.9233 - loss: 0.1881 - val_accuracy: 0.9681 - val_loss: 0.1077
Epoch 2/10
5956/5956 ━━━━━━━━━━━━━━━━━━━━ 22s 4ms/step - accuracy: 0.9685 - loss: 0.0858 - val_accuracy: 0.9738 - val_loss: 0.0928
Epoch 3/10
5956/5956 ━━━━━━━━━━━━━━━━━━━━ 43s 4ms/step - accuracy: 0.9755 - loss: 0.0696 - val_accuracy: 0.9559 - val_loss: 0.1307
Epoch 4/10
5956/5956 ━━━━━━━━━━━━━━━━━━━━ 23s 4ms/step - accuracy: 0.9796 - loss: 0.0598 - val_accuracy: 0.9808 - val_loss: 0.0807
Epoch 5/10
5956/5956 ━━━━━━━━━━━━━━━━━━━━ 26s 4ms/step - accuracy: 0.9817 - loss: 0.0535 - val_accuracy: 0.9781 - val_loss: 0.0842
Epoch 6/10
5956/5956 ━━━━━━━━━━━━━━━━━━━━ 25s 4ms/step - accuracy: 0.9845 - loss: 0.0464 - val_accuracy: 0.9777 - val_loss: 0.0876
Epoch 7/10
5956/5956 ━━━━━━━━━━━━━━━━━━━━ 24s 4ms/step - accuracy: 0.9856 - loss: 0.0457 - val_accuracy: 0.9687 - val_loss: 0.1035
Epoch 8/10
5956/5956 ━━━━━━━━━━━━━━━━━━━━ 22s 4